File: Demos/YMusic/src/Mgr/CacheMgr/CacheMgr.cpp
#include "stdafx.h"
#include "CacheMgr.h"
#include "Utility/Convert.h"


CCacheMgr::CCacheMgr(void)
{
	m_nTotalCacheSize = 0;
}

CCacheMgr::~CCacheMgr(void)
{
}

void CCacheMgr::Init(const DuiLib::CDuiString& sCachePath)
{
	DuiLib::CDuiString sTemp = sCachePath;
	sTemp.Replace('\\', '/');
	if(sTemp.GetAt(sTemp.GetLength()-1) != _T('/'))
	{
		sTemp += _T("/");
	}
	m_sCachePath = sTemp;
	m_sLrcPath = m_sCachePath + _T("lrc/");
	em_utility::mci::confirm_dir(m_sLrcPath.GetData(),_T('/'));
	m_sSongPath = m_sCachePath + _T("song/");
	em_utility::mci::confirm_dir(m_sSongPath.GetData(),_T('/'));
	m_sImagePath = m_sCachePath + _T("image/");
	em_utility::mci::confirm_dir(m_sImagePath.GetData(),_T('/'));
	m_sWebPath = m_sCachePath + _T("web/");
	em_utility::mci::confirm_dir(m_sWebPath.GetData(),_T('/'));
	m_nTotalCacheSize = GetFolderSize(m_sLrcPath) + GetFolderSize(m_sSongPath) + GetFolderSize(m_sImagePath) + GetFolderSize(m_sWebPath);
	StatFolderSize(0);
}

void CCacheMgr::ClearAllCatch()
{	
// 	CleanFolder(m_sLrcPath);
// 	CleanFolder(m_sSongPath);
// 	CleanFolder(m_sImagePath);
// 	CleanFolder(m_sWebPath);
	m_nTotalCacheSize = 0;
}

void CCacheMgr::CleanFolder(const DuiLib::CDuiString& sPath)
{
	WIN32_FIND_DATA FIND_DATA;
	int nTotalSize = 0;
	HANDLE h_find = INVALID_HANDLE_VALUE;

	DuiLib::CDuiString sFilePath = sPath + _T("/*.*");
	h_find = ::FindFirstFile(sFilePath,&FIND_DATA);
	if (h_find == INVALID_HANDLE_VALUE) 
		return ;

	do 
	{
		DuiLib::CDuiString s_file_name = sPath;
		s_file_name += FIND_DATA.cFileName;
		em_utility::mci::delete_file(s_file_name.GetData());
	} while (::FindNextFile(h_find, &FIND_DATA));
	FindClose(h_find);
}

size_t CCacheMgr::GetFolderSize(const DuiLib::CDuiString& sPath)
{
	WIN32_FIND_DATA FIND_DATA;
	size_t nTotalSize = 0;
	HANDLE h_find = INVALID_HANDLE_VALUE;

	DuiLib::CDuiString sFilePath = sPath + _T("*.*");
	h_find = ::FindFirstFile(sFilePath,&FIND_DATA);
	if (h_find == INVALID_HANDLE_VALUE) 
		return 0;

	do 
	{
		DuiLib::CDuiString s_file_name = sPath;
		s_file_name += FIND_DATA.cFileName;
		nTotalSize += em_utility::mci::get_file_size(s_file_name.GetData());
	} while (::FindNextFile(h_find, &FIND_DATA));

	FindClose(h_find);
	return nTotalSize;
}

size_t CCacheMgr::GetAllCatchSize()
{
	return m_nTotalCacheSize;
}

void CCacheMgr::StatFolderSize(size_t nNewFileSize)
{

		
}


DuiLib::CDuiString CCacheMgr::GetLrcPath()
{
	return m_sLrcPath;
}

DuiLib::CDuiString CCacheMgr::GetSongPath()
{
	return m_sSongPath;
}

DuiLib::CDuiString CCacheMgr::GetImagePath()
{
	return m_sImagePath;
}

DuiLib::CDuiString CCacheMgr::GetWebPath()
{	
	return m_sWebPath;
}

File: skin/ymusic/lua/main/CacheMgr.lua
local CCacheMgr = FLua.Class("CCacheMgr")

function CCacheMgr:__ctor()
	self.m_sCachePath = ""
	self.m_nTotalCacheSize = 0
	self.m_sLrcPath = ""
	self.m_sSongPath = ""
	self.m_sImagePath = ""
	self.m_sWebPath = ""
end
function CCacheMgr:Init(path)
	assert(Application.CreateDirectory(path))
	self.m_sCachePath = path:gsub("\\", "/")
	if self.m_sCachePath:sub(-1,-1) ~= "/" then
		self.m_sCachePath = self.m_sCachePath .. "/"
	end

	self.m_sLrcPath = self.m_sCachePath .. "lrc/"
	self.m_sSongPath = self.m_sCachePath .. "song/"
	self.m_sImagePath = self.m_sCachePath .. "image/"
	self.m_sWebPath = self.m_sCachePath .. "web/"

	self:StatFolderSize(0)
end

function CCacheMgr:GetCacheDir()
	return self.m_sCachePath
end

function CCacheMgr:GetAllCatchSize()
	return self.m_nTotalCacheSize
end

function CCacheMgr:ClearAllCatch()
	--self:CleanFolder(self.m_sLrcPath)
	--self:CleanFolder(self.m_sSongPath)
	--self:CleanFolder(self.m_sImagePath)
	--self:CleanFolder(self.m_sWebPath)
	self.m_nTotalCacheSize = 0
end

function CCacheMgr:StatSize(nNewFileSize)
end

function CCacheMgr:GetLrcPath()
	return self.m_sLrcPath
end

function CCacheMgr:GetSongPath()
	return self.m_sSongPath
end

function CCacheMgr:GetImagePath()
	return self.m_sImagePath
end

function CCacheMgr:GetWebPath()
	return self.m_sWebPath
end

--private:
function CCacheMgr:GetFolderSize(sPath)
	local strPath = sPath:gsub("\\", "/")
	if strPath:sub(-1,-1) ~= "/" then
		strPath = strPath .. "/"
	end
	local nTotalSize = 0
	local data = Application.FindFirstFile(strPath .. "*.*")
	local bStop = false
	while true do 
		if not data or bStop then
			break
		end

		local info = Application.GetFindData(data, strPath)
		local s_file_name = strPath .. info.cFileName
		nTotalSize = nTotalSize + info.st_size
		if not Application.FindNextFile(data) then
			bStop = true
		end
	end
	Application.FindClose(data)
	return nTotalSize
end
function CCacheMgr:CleanFolder(sPath)
	local strPath = sPath:gsub("\\", "/")
	if strPath:sub(-1,-1) ~= "/" then
		strPath = strPath .. "/"
	end

	local data = Application.FindFirstFile(strPath .. "*.*")
	local bStop = false
	while true do 
		if not data or bStop then
			break
		end

		local info = Application.GetFindData(data, strPath)
		local s_file_name = strPath .. info.cFileName
		os.remove(s_file_name)
		if not Application.FindNextFile(data) then
			bStop = true
		end
	end
	Application.FindClose(data)
end
function CCacheMgr:StatFolderSize(nNewFileSize)
end

return CCacheMgr